server/src/auth.rs

```diff
@@ -0,0 +1,119 @@
+use crate::config::Config;
+use crate::database::User;
+use futures::{future, Future};
+use lazy_static::lazy_static;
+use rand::RngCore;
+use tokio_threadpool::ThreadPool;
+use unicode_normalization::UnicodeNormalization;
+use vertex_common::{ServerError, UserId};
+
+lazy_static! {
+    static ref THREAD_POOL: ThreadPool = ThreadPool::new();
+}
+
+#[derive(Debug)]
+#[repr(u8)]
+pub enum HashSchemeVersion {
+    Argon2V1 = 1,
+}
+
+impl From<i16> for HashSchemeVersion {
+    fn from(v: i16) -> Self {
+        match v {
+            1 => HashSchemeVersion::Argon2V1,
+            _ => panic!("Invalid hash scheme version {}"),
+        }
+    }
+}
+
+pub fn valid_password(password: &str, config: &Config) -> bool {
+    password.len() <= config.max_password_len as usize
+        && password.len() >= config.min_password_len as usize
+}
+
+pub fn valid_display_name(display_name: &str, config: &Config) -> bool {
+    display_name.len() <= config.max_display_name_len as usize
+        && display_name.len() >= config.min_display_name_len as usize
+}
+
+fn valid_username(username: &str, config: &Config) -> bool {
+    username.len() <= config.max_username_len as usize
+        && username.len() >= config.min_username_len as usize
+}
+
+pub struct TooShort;
+
+pub fn process_username(username: &str, _config: &Config) -> String {
```

Comment:
`normalize_username`?

Reply:
Normalization is a unicode thing. Basically it makes the string into a consistent and comparable form. We use the most destructive form of normalization, where an anarchist circle-A will get converted to an A. The reason that we do this is to allow logins from different computers with perhaps different input languages (see what happened to Spotify), and also to make it more difficult to impersonate others using homoglyph attacks (e.g cyrillic a vs latin a).

Reply:
I was proposing to rename it. :P 
But thanks for the explanation! I understand it better now :slightly_smiling_face:

Reply:
Ah, I see. `process_username` also makes it lowercase, so it isn't just normalisation, though.